Revamp `RPC API For Lisk Node` page - Closes 1538 (#1611)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/api/lisk-node-rpc.adoc

= RPC API endpoints for Lisk nodes
Mona Bärenfänger <mona@lightcurve.io> Muhammad Talha <muhammad.talha@lightcurve.io>
// Settings
:toc: preamble
:toclevels: 5
:page-toclevels: 4
:idprefix: 0
:idseparator: -
// URLs
:docs_sdk: lisk-sdk::
// Project URLs
:url_general_endpoints: understand-blockchain/sdk/endpoints.adoc
:url_modules: {docs_sdk}modules/index.adoc
:url_plugins: {docs_sdk}plugins/index.adoc
:url_subscribe_event: {url_general_endpoints}#how-to-subscribe-to-events
:url_API_client: {url_general_endpoints}#the-api-client

A reference of all RPC endpoints and events is provided by the Lisk Framework for a blockchain application.

> TIP: For more information about the public, application endpoints, and events, check out the xref:{url_general_endpoints}[] page.

Author: Agreed, text updated

> == Module and plugin-specific endpoints and events

Author: Agreed, text updated.

> Modules and plugins often expose endpoints and events to the blockchain application. All endpoints and events registered with the application are available as RPC endpoints. For details about modules and plugins, see xref:{url_modules}[] and xref:{url_plugins}[].

Reviewer: ```suggestion
Modules and plugins often expose custom endpoints to the blockchain application, which then become available as RPC endpoints.
For details about endpoints for modules and plugins, see xref:{url_modules}[] and xref:{url_plugins}[].
```

== Usage

> Endpoints and events exposed by Lisk are categorized based on various namespaces. These namespaces point toward various, modules, plugins, and default application-specific endpoints and events. To invoke an endpoint or to subscribe to an event, combine the name of an endpoint/event with its namespace. For example:

Reviewer: ```suggestion
RPC endpoints and events exposed by a blockchain application are categorized based on various namespaces. These namespaces point toward various, modules, plugins, and default application-specific domains.
To invoke an endpoint or to subscribe to an event, combine the name of an endpoint/event with its namespace.
For example:
```

* `namespace_endpointName`
* `namespace_eventName`

For details, check out the <<endpoints, endpoints>> and <<events, events>> discussed below.

== Endpoints

Endpoints are invoked to fetch specific data for an application. Lisk node endpoints are public and can be invoked via RPCs. Lisk supports three modes of communication for invoking an endpoint as discussed in xref:{url_API_client}[The API client] section. The default RPC endpoints have been categorized based on various namespaces, as follows:

> . Chain

Reviewer: ```suggestion
. <<chain>>
```

adding internal anchors here might be beneficial

Author: Agreedl text updated

[source,js]
----
[
   {
      "ipAddress": "127.0.0.1",
      "port": 4532
      "networkIdentifier": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
      "networkVersion": "2.0",
      "nonce": "0fe9a3f1a21",
   }
]
----
====

==== `network_getStats`
Returns network statistics.

Request
[tabs]

=====
Specification::
+
--
Empty request body
--
cURL::
+
--
[source,json]
----
curl --location --request POST 'http://localhost:7887/rpc' \
--header 'Content-Type: application/json' \
--data-raw '{
    "jsonrpc": "2.0",
    "id": "1",
    "method": "network_getStats",
    "params": {}
}'
----
--
=====

.Response
[%collapsible]
====
.Example output
[source,js]
----
{
   "startTime": 1660638883613,
   "incoming": {
      "count": 0,
      "connects": 0,
      "disconnects": 0
   },
   "outgoing": {
      "count": 0,
      "connects": 0,
      "disconnects": 460
   },
   "banning": {
      "bannedPeers": {},
      "count": 0
   },
   "totalConnectedPeers": 0,
   "totalDisconnectedPeers": 0,
   "totalErrors": 454,
   "totalPeersDiscovered": 0,
   "totalRemovedPeers": 466,
   "totalMessagesReceived": {},
   "totalRequestsReceived": {}
}
----
====

== Events

> Events exposed by the Lisk RPC node can be subscribed as explained in the xref:{url_subscribe_event}[Subscribing to the events] section. To subscribe to an event, a WebSocket or Inter-Process Communication protocol must be used. The default events exposed by the Lisk Framework are described below:

Reviewer: Are these all events that are available? E.g. I think it was mentioned that every transactions triggers an event, announcing is the tx failed or was successful, is this included here?

Author: In total there are eight events as mentioned in the [events.ts file](https://github.com/LiskHQ/lisk-sdk/blob/development/framework/src/engine/events.ts). Due to the unavailability of some content, I have left a note [here](https://github.com/LiskHQ/lisk-docs/issues/1538) after discussing it with Shu

= RPC API for Lisk nodes
Mona Bärenfänger <mona@lightcurve.io> Muhammad Talha <muhammad.talha@lightcurve.io>
// Settings
:toc: preamble
:toclevels: 5
:page-toclevels: 4
:idprefix: 0
:idseparator: -
// URLs
:docs_sdk: lisk-sdk::
// Project URLs
:url_general_endpoints: understand-blockchain/sdk/endpoints.adoc
:url_modules: {docs_sdk}modules/index.adoc
:url_plugins: {docs_sdk}plugins/index.adoc
:url_subscribe_event: {url_general_endpoints}#how-to-subscribe-to-events
:url_API_client: {url_general_endpoints}#the-api-client

A reference of all RPC endpoints and events is provided by the Lisk Framework for a blockchain application.

TIP: For more information about RPC endpoints, events, and the Lisk communication architecture in general, check out the xref:{url_general_endpoints}[] page.

== Custom RPC endpoints for Modules and Plugins

Modules and plugins often expose custom endpoints to the blockchain application, which then become available as RPC endpoints.
For details about endpoints for modules and plugins, see xref:{url_modules}[] and xref:{url_plugins}[].

> == Usage

Reviewer: Aliases and usage examples are also described on the SDK RPC page updated in #1513 , please add links to the relevant sections at this place. Alternatively, you can update this section also in PR #1513, as you prefer :)

Author: Ummmm, various Links to the **API Client** and **Subscribing Event** sections are present in this PR. Also, I already updated the Alias section in PR #1513 so not sure about the suggested change there.  The best I can do here is add a tip about the API Client and Subscribing Events section, however, that would be redundant. What do you think?

Reviewer: Hmm, yeah I think the content under the heading `Usage` is maybe not really fitting here, and you are right it is a little redundant to the `aliases` section in #1513. So maybe we can remove the content here, or merge it with the `aliases` section.

On the other hand, the content in the next section `Endpoints` in line 40 explains the usage of the RPC API, and also provides a link to the usage example in #1513. I think we could move this up under the `usage` section, if we want to keep this section(except of the last sentence in line 40).

- Btw we normally use a new line for every new sentence, it makes overview and review easier :)

Author: Thanks, I have removed the `Usage` section as you suggested. I have also incorporated the new line convention in this file. Will follow it in the future as well :)

Request
[tabs]

=====
Specification::
+
--
Empty request body
--
cURL::
+
--
[source,json]
----
curl --location --request POST 'http://localhost:7887/rpc' \
--header 'Content-Type: application/json' \
--data-raw '{
    "jsonrpc": "2.0",
    "id": "1",
    "method": "network_getStats",
    "params": {}
}'
----
--
=====

.Response
[%collapsible]
====
.Example output
[source,js]
----
{
   "startTime": 1660638883613,
   "incoming": {
      "count": 0,
      "connects": 0,
      "disconnects": 0
   },
   "outgoing": {
      "count": 0,
      "connects": 0,
      "disconnects": 460
   },
   "banning": {
      "bannedPeers": {},
      "count": 0
   },
   "totalConnectedPeers": 0,
   "totalDisconnectedPeers": 0,
   "totalErrors": 454,
   "totalPeersDiscovered": 0,
   "totalRemovedPeers": 466,
   "totalMessagesReceived": {},
   "totalRequestsReceived": {}
}
----
====

== Events

Events exposed by the Lisk RPC node can be subscribed as explained in the xref:{url_subscribe_event}[subscribing to the events] section. To subscribe to an event, a WebSocket or Inter-Process Communication protocol must be used. The default events exposed by the Lisk Framework are described below:

> 

Reviewer: A list of the events namespaces, analog to the list for the endpoints at line 42-47 would be nice

Author: Done

[source,json]
----
{
   "transaction": {
   "id": "95b18ca901c910ea34d5df8896f6a9bc477f773ba5d0ff08c500711c15efb1db",
   ...
   }
}
----

> ====

Reviewer: Please delete the old content, instead of commenting out.

Author: Will do.
